### ISSUE TITLE
aws_db_instance from snapshot doesn't apply backup_retention_period or tags updates on 1st run

### ISSUE THREAD

**Issue by maxleonca** [terraform-providers/terraform-provider-aws#5959]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
Terraform v0.11.8
+ provider.aws v1.37.0
+ provider.null v1.0.0

### Affected Resource(s)

* aws_db_instance

### Terraform Configuration Files

```hcl
data "aws_db_snapshot" "db_prod_snapshot" {
  most_recent = true
  db_instance_identifier = "production-db"
}

resource "aws_db_instance" "staging_db" {

  identifier = "staging-vpc"

  snapshot_identifier = "${data.aws_db_snapshot.db_prod_snapshot.id}"
  skip_final_snapshot = true
  instance_class = "${var.db_instance_class}"
  auto_minor_version_upgrade = false
  apply_immediately = true

  db_subnet_group_name = "${var.db_subnet_group}"
  vpc_security_group_ids = "${var.rds_security_groups}"

  backup_retention_period = 0

  tags {
  }
}

resource "null_resource" "setup_staging" {
  depends_on = ["aws_db_instance.staging_db"]
  provisioner "local-exec" {
    command = "bash scripts/setup_db.sh"

  }
}
```

### Output
On 1st run:
Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
The other resource is a null_resource to execute a couple of SQL scripts, runs flawlessly

On 2nd run:
Terraform will perform the following actions:

  ~ aws_db_instance.staging_db
      backup_retention_period: "30" => "0"
      tags.%:                  "1" => "0"
      tags.mytag:               "" => ""

### Expected Behavior
Both the specified backup_retention_period and the update on the tags should be applied on the 1st run.

### Actual Behavior
It created the DB as specified but the modifications for backup_retention_period and tags are ignored unless the plan is applied again

### Steps to Reproduce
Run the example I provided, (just check the data for the snapshot and update accordingly).

1. `terraform apply`

### Important Factoids
Running inside a VPC

Thanks

**Comment by bflad** [comment]
Hi @maxleonca 👋 Thanks for reporting this and sorry you're running into trouble here.

So interestingly enough, I just recently went through and added testing with this resource for `snapshot_identifier` and many of the other arguments (mostly to ensure that they all properly call `ModifyDBInstance` after `RestoreDBInstanceFromSnapshot` if they need to), but only tested the "non-zero" value cases in the restore from a snapshot with zero values.

https://github.com/terraform-providers/terraform-provider-aws/blob/2cead4b6976081a831070de38c5be720c10ca56c/aws/resource_aws_db_instance_test.go#L815-L840

Seems like "unsetting" some of the snapshot details is a likely case too!  Can you confirm that the production database/snapshot in your case has `backup_retention_period` set to `30` and the tag the restored database is trying to remove? Thanks.

**Comment by bflad** [comment]
Work in progress pull request submitted: #5970 (handles `backup_retention_period` but not `tags` yet since the fix will be very different)

**Comment by maxleonca** [comment]
Yes indeed is the case, on the 2nd run retention_period goes from 30 to 0 and the tags get scratched.

**Comment by bflad** [comment]
The fix to properly apply `backup_retention_period = 0` on first apply has been merged and will release with version 1.38.0 of the AWS provider, likely later today. The `tags` portion will require a slightly different fix at some point.
